datastore/src/main/java/org/jboss/pnc/datastore/DefaultDatastore.java

```diff
@@ -214,11 +220,17 @@ private Set<Artifact> saveArtifacts(
 
         if (artifactConstraints.size() > 0) {
             logger.debug("Searching artifacts by {} constraints.", artifactConstraints.size());
-            List<Artifact> artifactsInDb = artifactRepository
-                    .queryWithPredicates(ArtifactPredicates.withIdentifierAndSha256(artifactConstraints));
-            for (Artifact artifact : artifactsInDb) {
-                logger.trace("Found in DB, adding to cache. Artifact {}", artifact);
-                artifactCache.put(artifact.getIdentifierSha256(), artifact);
+            // [NCLSUP-912] partition the constraints in maximum size to avoid a stackoverflow in Hibernate
+            // We use the Guava Lists.partition, which requires a List. Hence we have to convert it also
+            List<List<Artifact.IdentifierSha256>> partitionedList = Lists
+                    .partition(artifactConstraints.stream().collect(Collectors.toList()), QUERY_ARTIFACT_PARITION_SIZE);
+            for (List<Artifact.IdentifierSha256> partition : partitionedList) {
+                List<Artifact> artifactsInDb = artifactRepository
+                        .queryWithPredicates(ArtifactPredicates.withIdentifierAndSha256(partition));
+                for (Artifact artifact : artifactsInDb) {
+                    logger.trace("Found in DB, adding to cache. Artifact {}", artifact);
```

Comment:
would we want to trace with thousands of lines of artifacts?

Reply:
It was there before :) So there's no code change whether we partition it or not. Do you want me to remove it?

Reply:
right, keep it